### PR TITLE
[Bugfix] Programmatically updating layers won't toggle UI checkboxes

### PR DESCRIPTION
**Pull request by sisardor**

### CODE REVIEW

File: src/components/LayersPanel/LayersPanel.js [hunk 1]
```diff
@@ -27,6 +27,23 @@ class LayersPanel extends React.PureComponent {
...
+  componentWillUnmount() {
+    core.removeEventListener('layersUpdated', this.onLayersUpdated.bind(this));
+  }
+
+  onLayersUpdated() {
+    const doc = core.getDocument();
+    doc.getLayersArray().then(layers => {
+      if (layers.length) {
+        this.props.setLayers(layers);
+        this.forceUpdate();
```
Comment (ZhijieZhang) [inline src/components/LayersPanel/LayersPanel.js:43]: Can you explain why we need a `forceUpdate` here?

File: src/components/LayersPanel/LayersPanel.js [hunk 2]
```diff
@@ -27,6 +27,23 @@ class LayersPanel extends React.PureComponent {
...
   }
+  componentDidMount() {
+    core.addEventListener('layersUpdated', this.onLayersUpdated.bind(this));
+  }
+
+  componentWillUnmount() {
+    core.removeEventListener('layersUpdated', this.onLayersUpdated.bind(this));
+  }
+
+  onLayersUpdated() {
```
Comment (ZhijieZhang) [inline src/components/LayersPanel/LayersPanel.js:38]: Can we follow the convention in other components to use class properties?

```javascript
onLayersUpdated = () => {...}
```

And then we don't need to `.bind(this)`

File: src/components/LayersPanel/LayersPanel.js [hunk 3]
```diff
@@ -27,6 +27,23 @@ class LayersPanel extends React.PureComponent {
       window.docViewer.updateView();
     }
   }
+  componentDidMount() {
+    core.addEventListener('layersUpdated', this.onLayersUpdated.bind(this));
+  }
+
+  componentWillUnmount() {
+    core.removeEventListener('layersUpdated', this.onLayersUpdated.bind(this));
```
Comment (ZhijieZhang) [inline src/components/LayersPanel/LayersPanel.js:35]: It looks like this depends on a WebViewer PR, so let me know if  this is ready to merge.

**Comment by ZhijieZhang** [comment]
@sisardor do we still want this PR to be merged?

File: src/components/LayersPanel/LayersPanel.js (hunk 1 above)
Comment (sisardor) [reply src/components/LayersPanel/LayersPanel.js:43]: Because change is coming from backend, UI is not updating, so we need to force update

Comment (ZhijieZhang) [reply src/components/LayersPanel/LayersPanel.js:43]: hmm ok, so does `this.props.setLayers(layers);` actually do anything? Maybe we can add a comment explaining why `this.forceUpdate` is needed here.

Comment (sisardor) [reply src/components/LayersPanel/LayersPanel.js:45]: setLayers is updating parent component when layers change we supposed to update parent like in line 69.

File: src/components/LayersPanel/LayersPanel.js [hunk 4]
```diff
@@ -27,6 +27,25 @@ class LayersPanel extends React.PureComponent {
       window.docViewer.updateView();
     }
   }
+  componentDidMount() {
+    core.getDocument().on('layersUpdated', this.onLayersUpdated);
```
Comment (mparizeau-pdftron) [inline src/components/LayersPanel/LayersPanel.js:31]: It doesn't seem like this works as expected. componentDidMount is fired when the I actually open the panel so if there are any programmatic layer changes before the panel is opened then they won't be updated here.

Maybe we should be listening for this event somewhere else globally and calling the action there. Then in this component we'll just use the selector which also means we shouldn't need to use a forceUpdate?

**Comment by ZhijieZhang** [comment]
Closing as discussed at https://app.xodo.com/latest/#/team/-KEc2Okvm9KcebWPhOb7/-LFZ9idhkakrmXUrdxaO?msgId=-MFfWWpo6xB0rO9L5bHt

**Pull request closed without merging**